Lifecycle Methods and Event Alternative to `sessionsId` (#285)
ChainAgnostic/CAIPs: CAIPs/caip-x.md

---
caip: x
title: JSON-RPC Provider Lifecycle Methods for Session Management with CAIP-25 Sessions
author: [Alex Donesky] (@adonesky1)
discussions-to: TBD
status: Draft
type: Standard
created: 2024-06-07
requires: 25, 217
---

## Simple Summary

> CAIP-285 introduces new "lifecycle" methods and a new event for managing authorizations within an existing CAIP-25 session, allowing for addition, revocation, and retrieval of authorizations. These methods provide an alternative to session management via `sessionId`s, allowing `sessionId`s to be optional for CAIP-25.

Author: This file will be removed. Leaving it here for additional context for the time being

Reviewer: Counterargument: what if you left it up as an Implementer's Guide that explains the equivalence between legacy sessionId-managed sessions and transport-secured/single-session sessions?

## Abstract

> This proposal aims to extend the CAIP-25 standard by defining new JSON-RPC methods for managing the lifecycle of authorizations within a session. These methods allow dapps and wallets to dynamically adjust authorizations, providing more granular control and better user experience. Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.

Reviewer: ```suggestion
This backwards-compatible extension of the CAIP-25 standard defines new JSON-RPC methods for managing the lifecycle of authorizations within a session.
These methods allow dapps and wallets to dynamically adjust authorizations, providing more granular control and better user experience.
Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.
```

Author: done here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/405c5a03d7d0386c4f3d87bcf8db8de81af6257f

#### `provider_revoke`

Revokes permissions from an existing session.

**Parameters:**

- `sessionId` (string, optional): The session identifier.
- `scopes` (object, required): An object containing the scopes to be revoked, formatted according to CAIP-217.

**Initial Session Scopes:**

```json
{
  "eip155:1": {
    "methods": ["eth_signTransaction", "eth_sendTransaction"],
    "notifications": ["accountsChanged", "chainChanged"]
  },
  "eip155:137": {
    "methods": ["eth_sendTransaction"],
    "notifications": ["chainChanged"]
  }
}
```

**Example Request:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "provider_revoke",
  "params": {
    "scopes": {
      "eip155:1": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"]
      }
    }
  }
}
```

**Example Response:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "success": true,
    "updatedScopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction"],
        "notifications": ["accountsChanged"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"]
      }
    }
  }
}
```

**Explanation:**

> - The `provider_revoke` method removes the specified methods and notifications from the current session's permissions. If the scope no longer has any methods or notifications after revocation, it is removed from the session.

Reviewer 2: what if a session has no scopes left?

Author: Yeah we should call that out. I suppose the "session" is closed.

Reviewer 3: they could revoke `account` permissions in the session as well, right?

Author: yes for sure

Author: added some language to account for this and added `accounts` to the examples: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/51721fd888285fe5bba9bd3ca6d98929c73c5772

Reviewer 4: I think if no scopes are provided the method should treat it as a revocation of the whole session. No need to send the whole permission object over the wire if you can express intent in a more efficient manner. We can also of course continue to allow a caller to list the whole session if they so wish...

Author: Yeah that makes sense to me. Perhaps in the same vein sending an empty scopeObject revokes everything for that scope?

#### `provider_getSession`

Retrieves the current permissions of an existing session.

**Parameters:**

- `sessionId` (string, optional): The session identifier.

**Initial Session Scopes:**

```json
{
  "eip155:1": {
    "methods": ["eth_signTransaction"],
    "notifications": ["accountsChanged"]
  },
  "eip155:137": {
    "methods": ["eth_sendTransaction"],
    "notifications": ["chainChanged"]
  }
}
```

**Example Request:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "provider_getSession",
  "params": {}
}
```

**Example Response:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "scopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction"],
        "notifications": ["accountsChanged"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"]
      }
    }
  }
}
```

**Explanation:**

- The `provider_getSession` method returns the current permissions for the session. It lists all scopes along with their methods and notifications.

### Events

> #### `providerAuthorizationUpdated`

Reviewer 2: thoughts on `providerAuthorizationChanged`? matches the existing ethereum notifications

Author: That works!

Author: changed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/51721fd888285fe5bba9bd3ca6d98929c73c5772

---

> caip: XX

Reviewer: ```suggestion
caip: 285
```

## Abstract

This proposal aims to extend the CAIP-25 standard by defining new JSON-RPC methods for managing the lifecycle of permissions within a session. These methods allow dapps and wallets to dynamically adjust permissions, providing more granular control and better user experience. Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.

## Motivation

The motivation behind this proposal is to enhance the flexibility of CAIP-25 by enabling management of session permissions without sessionIds which don't map well to extension based wallet's dapp connections and could therefore add constraints and burdens to existing flows. The proposed methods provide an intuitive way to add, revoke, and retrieve permissions within an existing session, simplifying the management of session lifecycles.

### Use Case Scenarios

1. **Dapp Initiated Adding Permissions To an Existing Session:**

>    - **Current Method:** Call `provider_authorize` again with an existing session identifier and new scopes/methods to add. This is actually somewhat ambiguous in CAIP-25 where it’s unclear if incremental permissions adds should include the full object or just scopes to add: "This object gets updated, extended, closed, etc. by successive calls and notifications, each tagged by this identifier."

Reviewer: full object was intended but maybe not explicit enough in that sentence!

Author: Full object is ok but would be nice to have it be additive by default IMO?

Reviewer: i'm not against it, i just need (as ever!) to see the whole thing to compare to status quo, security considerations and all 😄

>    - **Proposed Method:** Use `provider_augment` to add new permissions to the existing session. Alternatively, this could be achieved with more specific language indicating that a subsequent `provider_authorize` call with new scopes/permissions could simply add to an existing session. If we pursue this latter approach one flow that will need clarity is what happens if a subsequent `provider_authorize` request includes `requiredScopes` which the respondent does not authorize. In this case I would expect that this does not revoke the existing session but rather only the new request is rejected.

Reviewer: that breaks semantics of required, tho? as-is, response can keep changing and wallet can keep responding with the same exact response to "reject changes"

Author: Not sure I follow 🤔

Author: I guess my preference for the request to only require the new "permissions" is that the dapp doesn't need to store or keep track of permissions that it already has. 

[At the same time part of me is leaning back towards keeping everything atomic](https://github.com/ChainAgnostic/CAIPs/pull/285#issuecomment-2198782292) and just saying either `provider_authorize`/`wallet_createSession` or `provider_augment`/`wallet_updateSession` should just request full object every time to change session scopes.

> 2. **Wallet Initiated Adding Permissions To an Existing Session:**

Reviewer 4: What's the use case here?

Author: This kindof thing:
https://github.com/ChainAgnostic/CAIPs/assets/34557516/bd22d855-ac36-40b4-ad4f-b186d4c060af

Reviewer: i don't understand

Author: In the video I add a permission/authorization for the site (in this case github) to use Account 3 all in the wallet (without any RPC Request Response cycle) and the site in this case (though it is not depicted) received an `accountsChanged` [event](https://docs.metamask.io/wallet/reference/provider-api/#accountschanged). This scenario `**Wallet Initiated Adding Permissions To an Existing Session:**` and the use of the `providerAuthorizationUpdated` would provide the equivalent functionality

Reviewer: So, just to be clear, is the idea here that the existing `accountsChanged` notification that some dapps already support, which can be authorized on eip155 `scopeObjects`, would trigger a change in the CAIP-25 config (i.e. add a member to the `chains` array in a `scopeObject` scoped to all of `eip155`), or that a net-new notification (mandatory to support for all CAIP-25 wallets on any chain) would replace it, or both? I can see pros and cons to all three, just want everything as explicit as possible for the sake of minimizing and tracking all possible breakage and corner cases.

This is currently possible connecting, say, MME over WC, but not AFAIK load-bearing in any dapps.

Author: So the purpose of this notification is to provide the wallet (or respondent) to modify (or signal the modification of) the session to the dapp (or caller). `accountsChanged` is limited to signaling to dApps that they have been granted permissions to new accounts. This event could be used to indicate any changes made by the user of the wallet to the permissions/authorizations that make up the session. The closest that CAIP-25 gets to providing this flow is:

> "If a respondent (e.g. a wallet) needs to initiate a new session, whether due to user input, security policy, or session expiry reasons, it can simply generate a new session identifier to signal this notification to the calling provider; if a caller needs to initiate a new session, it can do so by sending a new request without a sessionIdentifier. In such cases, a respondent (e.g. wallet) may choose to explicitly close all sessions upon generation of a new one from the same origin or identity, or leave it to time-out; maintaining concurrent sessions is discouraged (see Security Considerations)."

But in this case any modifications made on the wallet side would require tearing down the existing session and generating a new one - and it relies on sessionIds which we're looking to make optional. This doesn't seem very ergonomic or efficient. Let me know if I'm missing something.

   - **Current Method:** CAIP-25 does not make it very clear how a respondent (wallet), can modify the permissions of an existing session. The following excerpt is the closest we get: "The properties and authorization scopes that make up the session are expected to be persisted and tracked over time by both parties in a discrete data store, identified by an entropic identifier assigned in the initial response. This object gets updated, extended, closed, etc. by successive calls and notifications, each tagged by this identifier."

>    - **Proposed Method:** Wallet publishes and caller/dapp listens for an event `providerAuthorizationChanged` with the new full sessionScope.

Reviewer: i'd be ok with that, let's see what jxom and pedro think about its feasibility.  perhaps the name of the function should align with the EIP 6963 method/event names, to aid/boost adoption?

Author: I see this notification/event as less to do with discovery (a la 6963) and more to do with providing a path for wallets to communicate wallet-side changes to the session authorizations to the dapp.

3. **Wallet Initiated Permissions Revocation:**

   - **Current Method:** "If a respondent (e.g. a wallet) needs to initiate a new session, whether due to user input, security policy, or session expiry reasons, it can simply generate a new session identifier to signal this notification to the calling provider."
   - Given this language it is unclear if a wallet can revoke permissions without creating a new session. It is also therefore unclear if a wallet can revoke some subset of permissions without creating a new session.

>    - **Proposed Method:** Wallet publishes and caller/dapp listens for an event `providerAuthorizationChanged` with the new full sessionScope.

Reviewer 4: I'm assuming you're referring to user-initiated permission revocation in the extension? If so, what happens if the dapp is not connected to the wallet when the permission is revoked? Can the dapp just not rely on `provider_getSession`?

Author: > I'm assuming you're referring to user-initiated permission revocation in the extension? If so, what happens if the dapp is not connected to the wallet when the permission is revoked?

🤔 I guess you mean if the connection is severed somehow? I suppose in that case the dapp would use `provider_getSession` but I think this event is useful for most cases where a dapp can immediately reflect a change to the state of its permissions

Reviewer: severed, or just timed out.  i think the fundamental assumption that both app and wallet persist the config (if not the "session" in browser-memory or RPC-connection sense) is the first thing we need to be aligned on, before getting to other stuff?  in hassan's example, the wallet would let the user revoke a permission and... persist that message/event until there's an active RPC connection, right?

all this sounds fine to me, but should be explicit in the spec since it obligates the provider (or the wallet+plus+provider) to persist until next connection.  conversely, if this event is ONLY fired within an active session and dropped otherwise, specify that and separately define the behavior if user tells wallet to revoke without (or without knowing about the absence of) a live connection

Author: Does the following additional sentence assuage your concerns?

```
If a connection between the wallet and the caller/dapp is severed and the possiblity of missed events arises, the caller/dapp should immediately call `wallet_getSession` to retrieve the current session scopes.
```

4. **Dapp Initiated Permissions Revocation:**

>    - **Current Method:** "if a caller [dapp] needs to initiate a new session, it can do so by sending a new request without a sessionIdentifier."

Reviewer: ...on the assumption that dapp should only persist one "session" (config) at a time for any given wallet, thus new sessionId overwrites/replaces the revoked session, that by not being persisted is as good as dead.

(no change here, just filling in the assumptions/thought behind status quo in case it's cryptic)

   - **Proposed Method:** Use `provider_revoke` to revoke specific permissions or the entire existing session.

> 5. **Retrieving Current Permissions:**

Reviewer: ```suggestion
5. **Retrieving Current Authorizations:**
```
(Forgive the pedantry but permissions are at a higher layer in my personal mental model!)

>    - **Current Method:** Track permissions externally or infer from session creation.

Reviewer: ```suggestion
   - **Current Method:** Wallet and app both persist configuration across updates.
```
not sure what "infer from session creation" means?

Author: This rewording makes sense to me!

**Initial Session Scopes:**

```json
{
  "eip155:1": {
    "methods": ["eth_signTransaction"],
    "notifications": ["accountsChanged"],
    "accounts": ["eip155:1:0xabc123"]
  },
  "eip155:137": {
    "methods": ["eth_sendTransaction"],
    "notifications": ["chainChanged"],
    "accounts": ["eip155:137:0xdef456"]
  }
}
```

**Example Request:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "provider_augment",
  "params": {
    "scopes": {
      "eip155:1": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"]
      }
    }
  }
}
```

**Example Response:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "success": true,
    "updatedScopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction", "eth_sendTransaction"],
        "notifications": ["accountsChanged", "chainChanged"],
        "accounts": ["eip155:1:0xabc123"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"],
        "accounts": ["eip155:137:0xdef456"]
      }
    }
  }
}
```

**Explanation:**

> - The `provider_augment` method adds the specified scopes to the current session's permissions. If the scope already exists, the new methods and notifications are merged with the existing ones.

Reviewer: Why `updatedScopes` and not `sessionScopes`?

I'd specify more concretely, i.e.:

- if wallet rejects the augmentation, is the expected behavior returning the status-quo scopes or returning nothing or returning an explicit `no.` of some kind?

Author: `sessionScopes` works for me. I will also beef up the specification to include error codes and rejection flows.

**Initial Session Scopes:**

```json
{
  "eip155:1": {
    "methods": ["eth_signTransaction", "eth_sendTransaction"],
    "notifications": ["accountsChanged", "chainChanged"],
    "accounts": ["eip155:1:0xabc123", "eip155:1:0xdef456"]
  },
  "eip155:137": {
    "methods": ["eth_sendTransaction"],
    "notifications": ["chainChanged"],
    "accounts": ["eip155:137:0xdef456"]
  }
}
```

**Example Request:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "provider_revoke",
  "params": {
    "scopes": {
      "eip155:1": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"],
        "accounts": ["eip155:1:0xabc123"]
      }
    }
  }
}
```

**Example Response:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "success": true,
    "updatedScopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction"],
        "notifications": ["accountsChanged"],
        "accounts": ["eip155:1:0xdef456"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"],
        "accounts": ["eip155:137:0xdef456"]
      }
    }
  }
}
```

**Explanation:**

> - The `provider_revoke` method removes the specified methods, notifications, and accounts from the current session's permissions. If the scope no longer has any methods, notifications, or accounts after revocation, it is removed from the session. If no scopes remain in the session, the session is considered closed. If no accounts remain in a scope, and only write methods the entire scope is removed.

Reviewer: Either app or wallet can call this?
```suggestion
- The `provider_revoke` method removes the specified methods, notifications, and accounts from the current session's permissions. If a given `scopeObject` no longer has any methods, notifications, or accounts after executing this revocation, it is removed from the `sessionScopes` array. If no `scopeObject`s remain in the session after applying the above logic, the entire session is considered closed. If no accounts remain in a `scopeObject`, and no methods or only methods that would require an account to be called such as write methods, then the entire scope MUST be removed from the `sessionScopes` array of both wallet and application.
```

what does "considered closed" look like, behaviorally? Close the RPC session? Require a new _authorize call?  Reject all _augment calls until there is a new _authorize call?

Author: > Either app or wallet can call this?

Only dapp would call this.

**Parameters:**

- `sessionId` (string, optional): The session identifier.

**Initial Session Scopes:**

```json
{
  "eip155:1": {
    "methods": ["eth_signTransaction"],
    "notifications": ["accountsChanged"],
    "accounts": ["eip155:1:0xabc123"]
  },
  "eip155:137": {
    "methods": ["eth_sendTransaction"],
    "notifications": ["chainChanged"],
    "accounts": ["eip155:137:0xdef456"]
  }
}
```

**Example Request:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "provider_getSession",
  "params": {}
}
```

**Example Response:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "scopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction"],
        "notifications": ["accountsChanged"],
        "accounts": ["eip155:1:0xabc123"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": ["chainChanged"],
        "accounts": ["eip155:137:0xdef456"]
      }
    }
  }
}
```

**Explanation:**

- The `provider_getSession` method returns the current permissions for the session. It lists all scopes along with their methods, notifications, and accounts.

### Events

#### `providerAuthorizationChanged`

> This event is published by the wallet to notify the caller/dapp of updates to the session authorization scopes. The event payload indicates how the scopes have changed, showing additions and removals in the permissions.

Reviewer: Here I would actually prefer the whole config be sent rather than a delta/patch, for clarity and security.  Maybe a privacy wizard could convince me that's not a worthwhile tradeoff, but if nothing else, it makes feature parity with status quo easier to prove, since whole-configs get passed back and forth currently.

**Event Payload:**

- `sessionId` (string, optional): The session identifier.

> - `scopes` (object, required): An object containing the updated session scopes, formatted according to CAIP-217.

Reviewer: ```suggestion
- `sessionScopes` (array of `scopeObject` objects, required): An object containing the updated session scopes, formatted according to CAIP-217.
```

**Initial Session Scopes:**

```json
{
  "eip155:1": {
    "methods": ["eth_signTransaction"],
    "notifications": ["accountsChanged"],
    "accounts": ["eip155:1:0xabc123"]
  },
  "eip155:137": {
    "methods": ["eth_sendTransaction"],
    "notifications": ["chainChanged"],
    "accounts": ["eip155:137:0xdef456"]
  }
}
```

**Example Event:**

```json
{

>   "event": "providerAuthorizationChanged",

Reviewer 5: this should be a json-rpc notification

```
{
    "method": "providerAuthorizationChanged",
    "params": {
      ...
    }    
}
```

    "scopes": {
      "eip155:1": {
        "methods": ["eth_signTransaction", "eth_sendTransaction"],
        "notifications": ["accountsChanged"],
        "accounts": ["eip155:1:0xabc123"]
      },
      "eip155:137": {
        "methods": ["eth_sendTransaction"],
        "notifications": [],
        "accounts": ["eip155:137:0xdef456"]
      }
    }
  }
}
```

This event indicates how the scopes have changed by comparing the updated scopes with the initial session scopes. In the example, the method `eth_sendTransaction` was added to `eip155:1`, and the notification `chainChanged` was removed from `eip155:137`.

### Optional SessionIds

The `sessionId` parameter in the new lifecycle methods is optional. When not provided, the methods will operate on the current active session. This approach allows for more flexible session management without the overhead of tracking session identifiers.

## Security Considerations

The introduction of these lifecycle methods must ensure that only authorized parties can modify the permissions of a session. Proper authentication and authorization mechanisms must be in place to prevent unauthorized access or modifications.

## Privacy Considerations

> Managing permissions within an existing session reduces the need to create multiple session identifiers, which can help minimize the exposure of session-related metadata. However, care must be taken to handle these methods in a way that does not inadvertently leak sensitive information.

Reviewer: > Managing permissions within an existing session reduces the need to create multiple session identifiers
Here I feel like "session" would be more precisely worded "connection", and if it were, "multiple identifiers" wouldn't make any sense.  to put it another way, status quo is that wallet/provider persists the config of each connection and gives that config a connection-identifier which it only shares with the counterparty to that connection, so they aren't "multiple" unless provider/wallet is inadequately partioning its memory and storage between counterparties? 

I don't follow the bit about leaking information, since the current paradigm is to drop any messages not containing the counterparty-specific, opaque token and starting over at the first sign of mismatch between what you're persisting and what your counterparty is supposed to be persisting...

## Changelog

> - 2024-06-07: Initial draft of CAIP-XX.

Reviewer 6: ```suggestion
- 2024-06-07: Initial draft of CAIP-285.
```
